Quiz creation: Question replacement & a bevy of misc fixes and polish

nucleogenesis opened the pull request:
<!--
 1. Following guidance below, replace …'s with your own words
 2. After saving the PR, tick of completed checklist items
 3. Skip checklist items that are not applicable or not necessary
 4. Delete instruction/comment blocks
-->

## Summary
<!--
 * description of the change
 * manual verification steps performed
 * screenshots if the PR affects the UI
-->

- Adds the ability to replace selected questions
- Fixes #11826 
- Snackbars
- "Are you sure?" messaging and handling when closing a panel without saving changes
- Miscellaneous small fixes, as described in the commit list (I hope fairly clearly)

## References
<!--
 * references to related issues and PRs
 * links to mockups or specs for new features
 * links to the diffs for any dependency updates, e.g. in iceqube or the perseus plugin
-->

Closes #11017 
Closes #11013 
Closes #11015 
Closes #4708
Closes #4766

This directly impacts the following Product Issues - I hope that this can close them @radinamatic so please let me know your thoughts. I'll be going through them to make sure they're up-to-date with design decisions

- #10967 
- #10951 
- #10950 

## Reviewer guidance
<!--
 * how can a reviewer test these changes?
 * are there any risky areas that deserve extra testing
-->

### Known issues

_I will be preparing some follow-up issues for the items below. Sorry to disappoint re: the keyboard navigation @radinamatic 😭 but we'll get it worked out 💪🏻 _ 

- Whether or not a question's contents render is dependent on a Perseus bug [https://github.com/learningequality/kolibri/pull/9759] - if you go bottom-to-top then you'll see the questions as they'd look normally and can render multiple questions at once... but if you go top to bottom you'll see the bug.
- That same bug also means that the question replacement panel's accordion won't show a question's content unless you "Collapse all" of the items on the root page's accordion
- Sections drag & drop is a bit inconsistent
- ~Cannot navigate between the Section Tabs by way of keyboard due to [issue to be created soon]~ [FIXED](https://github.com/learningequality/kolibri/pull/11937/commits/cef5f454f3812504f6faf3200429dcb5eb41d4ae)


### QA Review

There are no gherkins yet for this version of the feature, so I suspect that walking through all possible paths is the best way to go. The issues in the 10xxx range are feature-level issues. 

I think that maybe I can lay out some potential variables that you can test combinations of to ensure things work as expected. I hope that this, in combination with the issues above, can serve as a guide of what to test.

#### Content on device

- When I have no content on my device and try to add resources
- When I have content on my device, but none of them include Exercises
- When I have any channels with Exercises 
- When I have Bookmarks, but none of them are Exercises
- When I have no Bookmarks
- When I have Bookmarks and some are Exercises

#### Side panel

- When I click outside the side panel
- When I do or do not have any changes made to the current page of the side panel do I get the "Are you sure?" message
- When I click the "back" arrow, do I go back to the page I was on before?
- Is the navigation intuitive?

#### Snackbars

- Do you get them when you think you should?

#### Content selection

- Import `tigil-fajod` for a channel that has one folder with enough exercises to trigger our lazy loading
- When you click "View more" at the bottom is the behavior as expected?

#### Add, Replace, Sort questions

- Can you make a quiz and do all the things?
- Sort
- Drag & drop
- Open close accordions
- Editing section
- Deleting sections
- Changing "question count" when deleting questions 
- Adding/removing questions to match changes to the "question count"

----

## Testing checklist

- [ ] Contributor has fully tested the PR manually
- [ ] If there are any front-end changes, before/after screenshots are included
- [ ] Critical user journeys are covered by Gherkin stories
- [ ] Critical and brittle code paths are covered by unit tests


## PR process

- [ ] PR has the correct target branch and milestone
- [ ] PR has 'needs review' or 'work-in-progress' label
- [ ] If PR is ready for review, a reviewer has been added. (Don't use 'Assignees')
- [ ] If this is an important user-facing change, PR or related issue has a 'changelog' label
- [ ] If this includes an internal dependency change, a link to the diff is provided

## Reviewer checklist

- Automated test coverage is satisfactory
- PR is fully functional
- PR has been tested for [accessibility regressions](http://kolibri-dev.readthedocs.io/en/develop/manual_testing.html#accessibility-a11y-testing)
- External dependency files were updated if necessary (`yarn` and `pip`)
- Documentation is updated
- Contributor is in AUTHORS.md

nucleogenesis committed:
update x of y replacements selected string

nucleogenesis committed:
remove unused style

nucleogenesis committed:
active & replacement question pools in useQuizCreation

nucleogenesis committed:
add counter_in_exercise to question titles in main accordion

nucleogenesis committed:
questions are replaced successfully & select all logic fixed

nucleogenesis committed:
fix checkbox logic on replacement list

nucleogenesis committed:
breadcrumbs polish

nucleogenesis committed:
channels link breadcrumbs works properly

nucleogenesis committed:
side panel doesn't open if the section_id in the URL is bad

nucleogenesis committed:
show questions in replacements accordions

nucleogenesis committed:
replace import/export lost in rebase

nucleogenesis committed:
SectionEditor styles tweaks

nucleogenesis committed:
checkbox cleanup in reploacement

nucleogenesis committed:
align "Add section" button

nucleogenesis committed:
misc style updates, fix duplicate back arrow in side panel

nucleogenesis committed:
fix delete section in section editor; add

nucleogenesis committed:
update & clean-up strings; replace ConfirmCancellationModal w/ KModals

nucleogenesis committed:
back arrow & x on resource selection

nucleogenesis committed:
basic general snackbar messages

nucleogenesis committed:
additional snackbar message

nucleogenesis committed:
ensure "no results" message shows when search yields nothing

nucleogenesis committed:
fix missing styles on bottom-nav of replace questions; snackbars for delete

nucleogenesis committed:
correct messaging for close confirmation select resources

nucleogenesis committed:
handle clicking outside of modal as closing side panel

nucleogenesis committed:
fix sidepanelmodal breakpoint for when it would go into mobile-mode

nucleogenesis committed:
fix number of resources on section editor

nucleogenesis committed:
fix browser nav issues w/ showBookmarks

nucleogenesis committed:
styles updates & message update

nucleogenesis committed:
spacing fix

nucleogenesis committed:
proper message on side panel back/close button

nucleogenesis committed:
outline-offset on section tabs fix

radinamatic commented:
I have yet to fire up the screenreader, but let's start with glitches in the keyboard navigation (testing with Firefox and Chrome on Ubuntu 20.04):

1. **Section order** seems to be outside of the `Tab` sequence, and no way to reorder the sections using the keyboard.
2. In the search results, pressing `Enter` on a resource card when in focus apparently exits the search results and returns to the full channel card. When the user does not try to search, but navigates through the folders until they get to the resource card pressing `Enter` on a resource card when in focus does nothing. If there is no way to actually **open the resource** to select single questions (whether by clicking with the mouse or pressing `Enter` when navigating by keyboard), the card should not be clickable and removed from the `Tab` sequence. The only actually _actionable_ element in this flow is the checkbox, correct @tomiwaoLE? Removing the cards themselves from the `Tab` sequence would make the keyboard navigation much easier...

1 | 2
-- | --
![section-order-no-tab](https://github.com/learningequality/kolibri/assets/1457929/ec933d55-f9f4-4053-a106-2cd5becf59be) | ![enter-on-resource-card-exits-search-results](https://github.com/learningequality/kolibri/assets/1457929/dc8503d5-912a-4e8d-92f3-1efe1925f316)

nucleogenesis commented:
Thanks Radina! The designs did not include an interface for selecting specific questions, although I do think that would be a relatively straightforward enhancement for someone to take on post-MVP. 

In the meantime, I will update this PR according to your feedback

nucleogenesis commented on kolibri/plugins/coach/assets/src/composables/useQuizCreation.js:
```diff
@@ -414,15 +418,27 @@ export default function useQuizCreation(DEBUG = false) {
+      pool.map(r => r.title),
+      pool.map(r => r.assessmentmetadata.assessment_item_ids),
+      get(_quiz).seed
+    );
```
I considered updating `selectQuestions` so that it took one object parameter but never came back to it. I think it might be worth doing to make reading calls to it easier to grok

rtibbles replied:
Assigning each pool.reduce and pool.map to intermediary variables might make this more readable?

The other thought that occurs is if `selectQuestions` is always operating on a `pool` object of known shape, couldn't it do this internally? Or are its other uses being invoked with other arguments?

nucleogenesis commented on kolibri/plugins/coach/assets/src/views/plan/CreateExamPage/CreateQuizSection.vue:
```diff
@@ -503,6 +511,13 @@
       },
     },
+    created() {
+      const { query } = this.$route;
+      if (query.snackbar) {
+        delete query.snackbar;
+        this.$router.replace({ query: { snackbar: null } });
+      }
```
I don't particularly like this solution but I had a hard time working out a way to show the snackbars during side-panel closing. I tried setTimout-ing it so that the snackbar could show before the panel closed but it was a weird pause between saying "close this" and seeing it closed. 

In any case, I'd appreciate thoughts on how this might be better approached.

rtibbles replied:
I might need more detail on the desired behaviour here to make a recommendation - is it that you are using the snackbar for confirmation before closing?

nucleogenesis replied:
The snackbar message like "Changes saved successfully" gets lost and doesn't show. So like

- In a side panel, click "Save"
- We call to show the snackbar at the end of the saving function
- The saving function also then calls to close the panel
- Closing the panel (or changing the route?) causes the snackbar not to show

So the solution here using the query params to pass the snackbar message around in this case feels kind of wack.

Like maybe I could try something like... provide a generic "showSnackbar" method from the quiz page root that we inject in child components so that maybe calling the function in the scope of the parent would allow us to show the snackbar while or after closing the side panel without having to set the query param.

Not sure if that would be more/less readable than this but I feel like this might be a bit brittle anyhow.

rtibbles replied:
So - I imagine this is happening because part of our coach global router hooks is to clear the snackbar?

https://github.com/learningequality/kolibri/blob/develop/kolibri/plugins/coach/assets/src/app.js#L52

So, maybe it would be simpler to add an exception here for this case?

rtibbles replied:
This might be a place where your use of nested routes might be an advantage - if the parent route hasn't changed, then don't trigger the clear.

nucleogenesis replied:
Omg the global router strikes again 😭 - thanks @rtibbles

nucleogenesis commented on kolibri/plugins/coach/assets/src/views/plan/CreateExamPage/CreateQuizSection.vue:
```diff
@@ -797,4 +820,8 @@
   }
 
+  /deep/ .tab {
+    outline-offset: -5px !important;
```
TODO: Add a comment explaining why this is here

nucleogenesis commented on kolibri/plugins/coach/assets/src/views/plan/CreateExamPage/ReplaceQuestions.vue:
```diff
@@ -1,35 +1,347 @@
+        updateSection,
       } = injectQuizCreation();
+
+      const showCloseConfirmation = ref(false);
+
+      function handleCancelClose() {
+        showCloseConfirmation.value = false;
+      }
+
+      function handleConfirmClose() {
+        context.emit('closePanel');
+      }
```
This ends up being repeated in each of the child components, once for each KModal and I wish I could think of a way to clean this up somehow. That or maybe if this is common to every use of KModal we could simplify this w/ a module in KDS that could make defining KModal's "show/hide/@confirm/@cancel" in a nicer way.

nucleogenesis commented on kolibri/plugins/coach/assets/src/views/plan/CreateExamPage/ReplaceQuestions.vue:
```diff
@@ -1,35 +1,347 @@
+    beforeRouteLeave(_, __, next) {
+      if (
+        !this.showCloseConfirmation &&
+        this.replacements.length &&
```
TODO: Explain this reasoning a bit as it is an edge case - we check for `this.replacements.length` because of a case where `this.replacements.length = this.selectedActiveQuestions.length = 0` -- and we only want to do this if there are no replacements happening.

rtibbles replied:
This feels like this should be a code comment? :)

nucleogenesis commented on kolibri/plugins/coach/assets/src/views/plan/CreateExamPage/ResourceSelection.vue:
```diff
@@ -112,11 +95,16 @@
-    <ConfirmCancellationModal
-      v-if="showConfirmationModal"
+    <KModal
+      v-if="showCloseConfirmation"
```
As I began using ConfirmCancellationModal I found it easier to work with to just use KModal directly rather than try to rework it to suit all of the use cases.

rtibbles replied:
That seems sensible! As I mentioned above, if it is feeling too verbose, defining handlers inline might help.

nucleogenesis commented on kolibri/plugins/coach/assets/src/views/plan/CreateExamPage/SectionSidePanel.vue:
```diff
@@ -16,43 +17,91 @@
+        }
+      }
+
+      onClickOutside(context.refs.resourcePanel, () => {
+        canCloseSidePanel.value = true;
+        handleClosePanel();
+      });
```
❤️ `@vueuse/core`  :) 

This made handling this particular case a breeze.

rtibbles commented on kolibri/plugins/coach/assets/src/composables/useQuizCreation.js:
```diff
@@ -414,15 +418,27 @@ export default function useQuizCreation(DEBUG = false) {
   /** @type {ComputedRef<QuizQuestion[]>} Questions in the active section's `resource_pool` that
    *                                         are not in `questions` */
-  const replacementQuestionPool = computed(() => {});
+  const replacementQuestionPool = computed(() => {
```
Sorry, I didn't spot this before, but this is important for data validity and avoiding weird bugs.

Code here seems to be assuming that `question_id` is a globally unique property of a question, whereas it is only guaranteed to be unique _within_ an exercise.

To specifically select a question, we need to specify both the question_id and the exercise_id - this is why for coach quizzes, we create a synthetic item id out of exercise_id and question_id: (nodeId and itemId) https://github.com/learningequality/kolibri/blob/develop/kolibri/plugins/learn/assets/src/views/ExamPage/index.vue#L391

We should use a similar synthetic identifier for checking whether questions have been selected or not.

Noting that this is a fairly constrained issue in terms of impact, as we are recording both the nodeId and the itemId as exercise_id and question_id in the exam data, so it's only for doing these crosschecks we need to be careful.

nucleogenesis replied:
Thanks Richard - to be sure I'm thinking about this correctly:

`exercise_id` - (AKA `ContentNode.id`?) - refers to any possible version of an Exercise resource and is consistent for that resource (aka ContentNode?)

`question_id` - (AKA `assessment_item_id`?) - refers to a unique question within a _particular version_ of the exercise and may be (or always will be?) different from previous versions of the same question

I'm not seeing exactly where using the `question_id` in this case results in a potential issue though for finding potential replacements so maybe I'm misunderstanding something above.

rtibbles replied:
Are we only allowing replacements from within the same content node? My assumption was that we can replace any question with any other question from the selected pool of exercises.

If this is the case, we might have selected two different exercises that have different contentnode ids, but have one or more assessment_item_ids that are the same, so that is where we could get an issue if we do not uniquely identify them by the combination of exercise_id and question_id.

nucleogenesis replied:
This has now penetrated into my brain for sure -- I think that perhaps I will make that into a follow-up issue, perhaps it can be well informed by a data architecture review as well!

rtibbles commented on kolibri/plugins/coach/assets/src/views/plan/CreateExamPage/ReplaceQuestions.vue:
```diff
@@ -1,35 +1,347 @@
+      :submitText="coreString('continueAction')"
+      :cancelText="coreString('cancelAction')"
+      :title="closeConfirmationTitle$()"
+      @cancel="handleCancelClose"
```
One possibility instead of defining methods for these is to just define the handlers inline:
```
@cancel="showCloseConfirmation = false"
```

This might make it feel a _little_ cleaner?

rtibbles commented on kolibri/plugins/coach/assets/src/views/plan/CreateExamPage/ReplaceQuestions.vue:
```diff
@@ -1,35 +1,347 @@
+
+      const replacements = ref([]);
+
+      function handleReplacement() {
```
It feels like some of this data manipulation might be more naturally handled inside the composable that's handling the question setting etc - maybe only the routing needs to happen here?

Not a blocker by any means, just trying to think through where everything lives.

nucleogenesis replied:
Great idea! Noted for a follow-up tech debt issue

rtibbles commented on kolibri/plugins/coach/assets/src/views/plan/CreateExamPage/ReplaceQuestions.vue:
```diff
@@ -1,35 +1,347 @@
+        });
+      }
+
+      function toggleInReplacements(question) {
```
As this is specific to the replacements that are tracked within this component, conversely, these seem to be more naturally here.

rtibbles commented on kolibri/plugins/coach/assets/src/views/plan/CreateExamPage/ReplaceQuestions.vue:
```diff
@@ -1,35 +1,347 @@
+
+      function toggleInReplacements(question) {
+        const replacementIds = replacements.value.map(q => q.question_id);
+        if (replacementIds.includes(question.question_id)) {
```
My point about the combined exercise_id and question_id seems to be applicable here too.

rtibbles reviewed (changes requested):
One blocker to do with ids - the rest is comment and questions.

marcellamaki commented:
Hi @nucleogenesis - I've started to do some QA here from the manual testing side of things, and I can see that you've laid out the guidance above, but I think having a little bit more clarity on "the scope of this PR is, 'does replacing a question work?'" vs. "this should be more or less feature completion, test away!" would be helpful, mostly because I'm finding some issues that aren't technical related to specifically replacing the quiz questions, but are definitely things that are broken or regressions (i.e. select a quiz no longer works). Would it be helpful for us to keep a narrow scoppe of QA review here to the question replacement specifically first and @radinamatic focusing on basic functionality and accessibility, merge that, and then do a bit of a step back and assess some potential other gaps, before the QA team really gets into testing the feature overall? I think there are some other quick win UI issues that we could open and put into the backlog for this upcoming iteration

nucleogenesis commented:
@marcellamaki thanks for that!

> Would it be helpful for us to keep a narrow scoppe of QA review here to the question replacement specifically first and @radinamatic focusing on basic functionality and accessibility, merge that, and then do a bit of a step back and assess some potential other gaps, before the QA team really gets into testing the feature overall? 

Yes please - the changes here are all specific around the feature of "Creating the quiz". If you happen to find any other issues or regressions, however, please make note of them here and they can be non-blocking in this PR context but I'll make issues out of them

nucleogenesis committed:
move exported method inline

nucleogenesis committed:
make leaf content nodes not clickable or focusable

nucleogenesis committed:
move cursor CSS change to /deep/ selector

nucleogenesis committed:
include `aria-selected` indicator on accordion-item and content card when item is selected

nucleogenesis committed:
comment on deep selector for section tabs

nucleogenesis committed:
comment clarifying ReplaceQuestions#beforeRouteLeave

radinamatic commented:
When navigating by keyboard inside the **Select folders or exercises from these channels** sidebar, `View more` button is not fully visible when in focus on Firefox (looks cut out as the sidebar is not fully scrolled down, looks OK in Chrome).

Firefox | Chrome
-- | --
<img width="639" alt="2024-03-08_17-17-08" src="https://github.com/learningequality/kolibri/assets/1457929/4070a6fe-42ca-4fb1-974d-605199f6bc4d"> | <img width="674" alt="2024-03-08_17-20-02" src="https://github.com/learningequality/kolibri/assets/1457929/02a5db27-24f0-41ac-b9df-e088add4bbef">

nucleogenesis committed:
section tabs have tabindex set to 0; keyboad focus fix from options dropdown

pcenov commented:
Hi @nucleogenesis - great to see this new feature!

Here are some observations mainly around adding/replacing questions:

| Issue| Screenshot |
:-------------------------:|:-------------------------:
| Replace questions - It shouldn't be possible to click the Replace button with no questions selected  | https://github.com/learningequality/kolibri/assets/79847249/47106903-f088-443a-98f1-72d562c9629c |
:-------------------------:|:-------------------------:
| Replace questions - The 'Select all' option is not working  | https://github.com/learningequality/kolibri/assets/79847249/f9fc4153-a5c3-4ec8-864e-cc877b720b1a |
:-------------------------:|:-------------------------:
| Replace questions - A "Changes saved successfully" message is shown when I close the overlay without making any changes | https://github.com/learningequality/kolibri/assets/79847249/2a69f01d-89c4-4a6c-89fe-5b0524ec4273 |
:-------------------------:|:-------------------------:
| Replace questions - Attempting to replace some questions with less or more than the selected questions should result in displaying the 'Review selected resources' modal (as per Figma designs), while currently it shows 'Are you sure you want to leave this page?' | https://github.com/learningequality/kolibri/assets/79847249/8b0c7452-234a-4224-8d1a-08186cde9c5f |
:-------------------------:|:-------------------------:
| Replace questions - Missing confirmation modal (as per Figma designs) when replacing questions | https://github.com/learningequality/kolibri/assets/79847249/a14eeb77-9092-4aaf-b28f-1cc98df10137 |
:-------------------------:|:-------------------------:
| Replace questions - Perhaps we shouldn't be showing bookmarked resources which are not exercises as it can be confusing | https://github.com/learningequality/kolibri/assets/79847249/ff4f3357-4d6c-40a8-b34c-f3bd573e2442 |
:-------------------------:|:-------------------------:
| Session settings - Getting an ''Are you sure you want to leave this page?' when I click on 'Change resources' | https://github.com/learningequality/kolibri/assets/79847249/640dc121-8df3-4764-b540-3b1d84af04f7 |
:-------------------------:|:-------------------------:

I noticed some other issues but since I am not 100% sure which ones are in scope for this task, let's start with this list.

In 'Session settings' the 'Number of questions' is set by default to 10 and as far as I can see the only way to add additional questions is by modifying that number (the -+ buttons are not always working correctly at the moment). So I am flagging that from a user's point of view it's not very clear how to add more questions to a section as the 'Add questions' button has disappeared.
Also flagging that the 'Replace' button looks like the browser's 'Refresh' button.

nucleogenesis committed:
move 'can we try to replace' logic to computed prop, disable Replace button

nucleogenesis committed:
fix & improve replacements KCheckbox handling

nucleogenesis committed:
fix checkbox console errors; allQuestionsSelected now returns Boolean, not int

bjester commented on kolibri/plugins/coach/assets/src/composables/useQuizCreation.js:
```diff
@@ -431,11 +447,14 @@ export default function useQuizCreation(DEBUG = false) {
+        get(activeQuestions)
+          .map(q => q.question_id)
+          .sort()
+      )
```
Would this perform any better than this implementation? seems like it avoids the sorts so it would minimize loops
```js
for (let activeQ of get(activeQuestions)) {
  if (!selectedActiveQuestions.includes(activeQ.question_id)) {
    return false;
  }
}
return true;
```

nucleogenesis replied:
You're has to be better for sure thanks for this @bjester !

rtibbles commented on test.html:
```diff
@@ -0,0 +1,107 @@
+<div
```
Wherever this should be, the root of the repo seems like the wrong place :)

nucleogenesis replied:
I was comparing the HTML output of two separate instances of KTabsList and trying to figure out why they'd be acting differently 😅 ... blast `git add .`

nucleogenesis committed:
accidental merge removed

nucleogenesis commented on kolibri/plugins/coach/assets/src/views/plan/CreateExamPage/CreateQuizSection.vue:
```diff
@@ -122,7 +123,8 @@
                 :hasIcons="true"
                 :options="activeSectionActions"
-                @tab="e => (e.preventDefault() || $refs.selectAllCheckbox.focus())"
+                @tab="$refs.addQuestionsButton.$el.focus()"
+                @close="$refs.addQuestionsButton.$el.focus()"
```
Why @tab & @close both doing the same?

- When closing by any means other than selecting an option, need to direct focus.

- Need to verify that `@tab` should close it?

nucleogenesis commented on kolibri/plugins/coach/assets/src/views/plan/CreateExamPage/CreateQuizSection.vue:
```diff
@@ -270,7 +281,7 @@
                           @click="toggleItemState(question.question_id)"
                         >
-                          <span>{{ title }}</span>
+                          <span>{{ title + " " + question.counter_in_exercise }}</span>
```
Should question.counter_in_exercise be wrapped in ()?

rtibbles replied:
I think that would be consistent with the previous implementation?

nucleogenesis committed:
move bookmarks exercise filter to where the value is set

this ensures we don't have to worry about filtering the list elsewhere
and fixes an issue where the # of bookmarks shown was ALL bookmarks,
not just exercises

nucleogenesis committed:
confirm replacement message

nucleogenesis committed:
don't ask for confirmation when going to select-resources from section editor

nucleogenesis committed:
selectQuestions new param "excludes" to avoid selecting questions already in section

nucleogenesis commented:
@pcenov I've pushed fixes for all of your noted issues except for two, which will be addressed in follow-up issues (or may not be doable...)

1) The issue w/ the "Changes saved" snackbar will be fixed in a new issue

2) The "you didn't select enough resources, try again" modal is impossible to trigger so long as we're disabling the "Replace" button unless the same conditions are met so I'll discuss this w/ @tomiwaoLE  and follow-up on this.

nucleogenesis committed:
clarify/clean up calls to selectQuestions

nucleogenesis committed:
use a unique id in place of question_id throughout quiz creation

marcellamaki reviewed (commented):
Hi @nucleogenesis thanks for all the work on this! I haven't done a thorough code review since the walk through last week, but I can if that would be helpful. 

I have done a little bit of manual QA this morning. I found some things that are outside of the scope of specifically _replacing questions_, but would be necessary to have done before we tag an alpha. (I suspect @pcenov and @radinamatic may have noticed them too, but are focusing specifically on the question replacement here). Namely, saving a quiz.

There are some errors with saving, and some improved validation on saving would also be important. Here's a couple of scenario's I've tried. 

- Not adding a title to the quiz: you get a 400 error, but there's no front end indication of what's wrong on the form (i.e. KTextbox error or helper text)

- Trying to save a quiz that has an empty section (no error message in the UI)
<img width="585" alt="Screenshot 2024-03-21 at 9 45 16 AM" src="https://github.com/learningequality/kolibri/assets/17235236/cca38808-a673-480f-ad42-cd3d5e402d53">

- Navigating a section where there are no questions added yet, or not the same number of questions in the section that matches the "Num questions" in the section settings.
<img width="625" alt="Screenshot 2024-03-21 at 9 48 11 AM" src="https://github.com/learningequality/kolibri/assets/17235236/bd17db4b-eb02-491d-b076-b6760c66e5b4">
(This might be worth talking to Tomiwa about because while it's existing functionality that we have a default of 10 questions, that can then be adjusted, in this new UI, it's hidden from the user and I think therefore more likely to have problems or mis-matches)

- Trying to save a quiz with no "apparent" issues (i.e. missing or empty fields)
<img width="504" alt="Screenshot 2024-03-21 at 9 55 23 AM" src="https://github.com/learningequality/kolibri/assets/17235236/096411bb-2330-4869-b3d5-26cad32e1c32">

The 400 error suggests that maybe the `validateQuiz()` is not catching everything, and the `id` error suggests that something is not quite matching up with the mapping. What I haven't dug deeply enough to figure out yet is if this is just a small bug with the function, or if there is something more foundational to the question tracking and replacement that needs to be updated.

All in all, I haven't been able to save a quiz. I realize that this code was committed a while ago, so it's fine with me to have this addressed in follow up - and I'm happy to open an issue! I just want Radina and Peter to be able to go through the whole flow when we start tagging pre-releases. :) They will be much happier if they can.

Non blocking but a pet peeve, lots of console warnings (I will not blame you for the KResponsiveWindow one as it is not your problem 😄): 
```
[WARN: kolibri/core/assets/src/router.js] Route name 'QUIZ_SELECT_RESOURCES' is not registered
```

Let me know what would be the most helpful in sorting out what to do now vs. what to do in follow up, and if you'd like me to open an issue.

pcenov commented:
Hi @marcellamaki indeed we are aware of the issues you have outlined above and we're disregarding them as known issues given the fact that the entire feature is a work in progress. Ideally at some point it will be possible to save and edit a quiz and then take the quiz as a learner plus having the data for the quiz results - looking forward to it! 

@nucleogenesis - I confirm that the issues I had previously reported are fixed now (excluding the 2 that you've commented on).

In today's round of testing I've been able to identify the following issues which I will list here, but which can be worked on separately if not in scope for this specific PR:

1. Section settings - Missing 'Select resources from channels' button when viewing the 'Section settings' for the first time. Currently I am seeing the 'Change resources' button which I would guess should be displayed only after the user has selected resources:
![2024-03-21_16-34-02](https://github.com/learningequality/kolibri/assets/79847249/bc776090-f371-4b19-bc13-966378f6e486)

2. Section settings - Once I delete all the questions I can't add them again as the 'Save changes' button remains disabled:

![2024-03-21_18-40-56](https://github.com/learningequality/kolibri/assets/79847249/a8dbf4e2-48aa-4528-bf22-2b4bfd250363)

3. Sections - If I have several sections when I delete the questions in section 2 and go to section 1 and then back to section 2 I can see that the questions have reappeared.
4. Sections - Changes made to the question order in a section are not saved.
5. Sections - The expand/collapse buttons all are not functioning correctly when switching between the sections. 
6. Sections - it's possible to add unlimited number of sections but once they get listed under the ellipsis button then there's no way to know which section I am viewing at the moment

![2024-03-21_19-05-13](https://github.com/learningequality/kolibri/assets/79847249/6d921571-7676-451e-a32d-0016938a26c2)

7. Sections - There's no confirmation prompt when deleting a section from the options menu (while there is such if I am deleting it

8. Questions - The question previews are not working at all now both in the question list and in 'Replace questions'. I know that this is partially a known issue, but previously when I tested I was seeing the previews so noting it here.

![2024-03-21_19-00-57](https://github.com/learningequality/kolibri/assets/79847249/197f60ea-8419-4608-a546-fa3a7b628189)

9. Number of questions - When I add let's say 10 questions and then I delete some of them and then I open the Section settings it incorrectly shows the default number of 10 questions which then results in issues when increasing or decreasing the number of questions.

nucleogenesis committed:
update delimiter for unique question id

nucleogenesis committed:
fixes for questions management edge cases; fix showing preview (as best as possible for now)

also the `counter_in_exercise` is now associated w/ the question's index in the exercise's assessment ids list

nucleogenesis committed:
conditional label for resource change button in section editor

nucleogenesis committed:
update questions more reliably

pcenov commented:
Hi @nucleogenesis the following issues from above are still not fixed:

1. Sections - The expand/collapse all buttons are not functioning correctly when switching between the sections.

https://github.com/learningequality/kolibri/assets/79847249/9073c2b5-7841-4a7f-84dd-1244eb70ec11

2. Sections - It's possible to add unlimited number of sections but once they get listed under the ellipsis button then there's no way to know which section I am viewing at the moment

https://github.com/learningequality/kolibri/assets/79847249/b25095aa-dbea-426e-a29a-1ec320d8cb7c

3. Sections - There's no confirmation prompt when deleting a section from the options menu (while there is such if I am deleting it

https://github.com/learningequality/kolibri/assets/79847249/c56976c9-326a-443f-903b-6e0ca2ceb6bd

Also now I can't delete the last remaining question in a section, so I can 't fully retest the issue "Section settings - Once I delete all the questions I can't add them again as the 'Save changes' button remains disabled":

https://github.com/learningequality/kolibri/assets/79847249/78732177-ce72-43ba-a722-543a3d013e52

New issues noticed:

1. Replace questions - I can't preview questions after having just replaced them:

https://github.com/learningequality/kolibri/assets/79847249/b652dd93-8477-47db-9649-2a7d8e0b04a9

2. Section settings - getting a "Changes saved successfully" message each time after just clicking outside of the modal:

https://github.com/learningequality/kolibri/assets/79847249/4a8f1114-31fc-4337-8479-77ab64227002

nucleogenesis committed:
update homehydrate viewset to account for Exam v3

nucleogenesis committed:
use unique_id_arrays while generating list of activequestions

nucleogenesis commented:
Thank you @pcenov , listing each issue noted:

**Previous issues**
- Unnecessary snackbar messages when closing side panel -- new issue for fixing Snackbars #12038 
- Identifying the active section -- to be fixed in #12011
- No confirmation when deleting section from Options dropdown -- tracked in #12037
- *Can't delete the final question* -- so the way I've worked this out is that a section cannot have 0 questions, so when you delete all of the questions, I default it to have a "question count" of 1. So in the case where you would have deleted all of the questions from the section, the only way to add questions back to it is to update the section's "question count". When you go to "Change resources" -- that's just updating the pool of questions from which we randomly select (also from which you can replace) within that particular section. I think it's worth considering an updated UX overall for this as it's kind of unclear what is happening when you delete questions (cc @tomiwao)
- "Collapse/Expand" state not being tracked while navigating between sections -- we've not set this up to track the collapsed/expanded state of the questions and each time you move to a new section it was intended during dev that they all start collapsed. This shouldn't be a blocker for merge, but I will discuss w/ @tomiwao

 **New issues**
- "Cannot preview questions after replacement" -- I've fixed this in a new commit

nucleogenesis committed:
remove unnecessary select questions test

The test here was previously accounting for a possibility where the same resource from a different channel was being used -- however, we only select questions now by using a fully unique ID combining the exercise_id:question_id so this test is no longer needed

nucleogenesis committed:
use proper node id for exercise_id, proper data_model property in examViewer module

nucleogenesis committed:
use exam_assignment_lookup for cleaner itearator

rtibbles commented on kolibri/plugins/coach/assets/src/composables/useQuizCreation.js:
```diff
@@ -156,9 +156,32 @@ export default function useQuizCreation(DEBUG = false) {
-    const { question_count } = updates;
-    if (question_count) {
-      if (question_count < (targetSection.question_count || 0)) {
+    // Make sure all questions in the an updated pool have proper unique IDs
```
Possible additional refactoring in the future - the resource pool and questions could maybe be factored out into separate functions, that are both called from this one?

rtibbles commented on kolibri/plugins/learn/viewsets.py:
```diff
@@ -182,15 +183,30 @@ def consolidate(self, items, queryset):
+        section_questions = list()
+
         for exam in exams:
-            exam_node_ids |= {
```
I mentioned this on slack, but this diff would be simpler:
```
from kolibri.core.exam.models import exam_assignment_lookup
```
then here:

```
            exam_node_ids |= {
                exercise_id for exercise_id, _ in exam_assignment_lookup(exam.get("question_sources", []))
            }
```

nucleogenesis replied:
Aye I hadn't considered that `exam_assignment_lookup` handles the backward compatibility for us , thanks!

rtibbles commented on kolibri/plugins/learn/viewsets.py:
```diff
@@ -218,7 +234,7 @@ def consolidate(self, items, queryset):
             exam["missing_resource"] = any(
                 question["exercise_id"] not in available_exam_ids
-                for question in exam.get("question_sources")
+                for question in (section_questions or exam.get("question_sources"))
```
and here:
```
for question, _ in exam_assignment_lookup(exam.get("question_sources", [])
```

This is better, because currently the `section_questions` will be collecting _all_ questions from every exam, so this will mean that if any quiz has missing resources, then all quizzes will be marked as having missing resources (because section_questions will always be all of the questions from every exam).

rtibbles reviewed (changes requested):
More than just being neater, I think there is an actual bug in the updates to the learn viewset.

Some tests of the useQuizCreation updates for the synthetic id generation would be helpful. I am also not seeing where we are stripping out those ideas before sending the data to the backend, but maybe that was already happening and I'm not seeing in it in a more restricted diff.

nucleogenesis committed:
use exam_assignment_lookup throughout as it is itself backwardcompatible

nucleogenesis committed:
remove outdated/unnecessary DEBUG functionality from useQuizCreation

nucleogenesis committed:
update quiz creation specs to clarify exercise_id origin

nucleogenesis committed:
add unique_question_ids to resources when they're saved in the first place

nucleogenesis committed:
properly handle resource_pool, question and question_count updates

this was kind of a pain to work out, ultimately, the method to select random questions
needed to take a parameter for the resource_pool it will select from.
this implementation results in a massive updateSection method. an ideal alternative
would probably involve changes to the API overall, which would require a moderate refactor
but would also likely result in a much cleaner interface

nucleogenesis committed:
better definition for QuizQuestion.id in specs

nucleogenesis committed:
update useQuizCreation spec data to accommodate unique IDs

pcenov commented:
Hi @nucleogenesis, here are the latest findings of my testing which I'll just list again for you to sort out other whether they will be addressed here or as follow-ups:

Questions replacement/deletion/order:

1. Replace questions - Previewing questions in the 'Replace questions' modal is not working. I'm guessing this is caused by the Perseus bug that you've already mentioned as a known issue, but since previewing the questions is at least partially working in the 'Question list' I'm mentioning it here as well:

https://github.com/learningequality/kolibri/assets/79847249/cece2e86-b617-4735-9ade-e00b0f0127bf

2. Replace questions - If I select one question and I replace it with another one, then the new question goes to the bottom of the question list, while ideally it should be displayed at the same position as the original question:

https://github.com/learningequality/kolibri/assets/79847249/735f9ff0-c2b1-45c1-8469-2dc4fed5f4ca

3. Question list - A question selected for replacement/deletion is being selected in several sections, while it should be selected only in the section that I am currently editing:

https://github.com/learningequality/kolibri/assets/79847249/ef067eba-f927-496a-aa69-a76a34a829f3

4. Question order - When I go to the section options, the question order is set to 'Fixed' by default. However if I select 'Randomized' and save the change, then the option to drag and reorder the question should no longer be available:

https://github.com/learningequality/kolibri/assets/79847249/dfe3ffe9-64a2-4043-abd2-c28a235aafb0

Other general issues:

1. Section settings - I can save a section without a title and questions:

https://github.com/learningequality/kolibri/assets/79847249/9359f6ad-9524-4693-8c1b-f3b683e3f2b7

2. Section settings - Not getting any notification that there are not enough questions if I attempt to add too many. In general the users should be notified what is the max number that they can add:

https://github.com/learningequality/kolibri/assets/79847249/0d8e31aa-c5cf-4438-b206-4c394e104397

3. Section settings - The scrollbar is not functioning correctly (there's none in Firefox) and the sections are displayed under the 'Delete section' and 'Apply settings' buttons:

https://github.com/learningequality/kolibri/assets/79847249/8c8bf703-adc3-4571-9a50-ea7fea886f66

nucleogenesis committed:
update og to original

nucleogenesis committed:
fix replacement contentrenderer id refs

nucleogenesis commented:
Thank you @pcenov ! <3

I fixed the issue w/ replacement questions previews.

The other items will be added to #12037 or discussed w/ designers next week, after which I'll make more issues to cover everything.

pcenov commented:
Thanks @nucleogenesis - I can confirm that in the latest build I can see the preview of the replacement questions but only if the questions in the 'Question list' are not expanded and if I am opening the preview separately for each question - in any case it's best for us to wait for the Perseus bug to be fixed to fully test the preview:

https://github.com/learningequality/kolibri/assets/79847249/5f1d268c-3a38-4bb3-9ac6-9b97af405f79

rtibbles commented on kolibri/plugins/coach/assets/src/views/plan/CreateExamPage/ReplaceQuestions.vue:
```diff
@@ -78,12 +78,12 @@
               :style="{ userSelect: dragActive ? 'none!important' : 'text' }"
             >
               <ContentRenderer
-                :ref="`contentRenderer-${question.id}`"
```
It might still be necessary to have this be `id` rather than `question_id` to prevent  a ref collision?

rtibbles reviewed (commented):
Just one last possible thing that I think might bite us.

nucleogenesis committed:
use question.id on ref to ensure uniqueness

rtibbles reviewed (approved):
This looks good to me - still some cleanup to do, but have been filed in follow up issues.
